packages/discord.js/src/managers/GuildMemberManager.js

```diff
@@ -246,46 +247,64 @@ class GuildMemberManager extends CachedManager {
     const query = initialQuery ?? (users ? undefined : '');
 
     return new Promise((resolve, reject) => {
-      this.guild.client.ws.send(this.guild.shardId, {
-        op: GatewayOpcodes.RequestGuildMembers,
-        // eslint-disable-next-line id-length
-        d: {
-          guild_id: this.guild.id,
-          presences,
-          user_ids: users,
-          query,
-          nonce,
-          limit,
-        },
-      });
       const fetchedMembers = new Collection();
       let index = 0;
+
+      const timeout = setTimeout(() => {
+        // eslint-disable-next-line no-use-before-define
+        this.client.ws.off(WebSocketShardEvents.Dispatch, rateLimitHandler);
+        // eslint-disable-next-line no-use-before-define
+        this.client.removeListener(Events.GuildMembersChunk, handler);
+        this.client.decrementMaxListeners();
+        reject(new DiscordjsError(ErrorCodes.GuildMembersTimeout));
+      }, time).unref();
+
       const handler = (members, _, chunk) => {
         if (chunk.nonce !== nonce) return;
 
-        // eslint-disable-next-line no-use-before-define
         timeout.refresh();
         index++;
         for (const member of members.values()) {
           fetchedMembers.set(member.id, member);
         }
 
         if (members.size < 1_000 || (limit && fetchedMembers.size >= limit) || index === chunk.count) {
-          // eslint-disable-next-line no-use-before-define
           clearTimeout(timeout);
+          // eslint-disable-next-line no-use-before-define
+          this.client.ws.off(WebSocketShardEvents.Dispatch, rateLimitHandler);
           this.client.removeListener(Events.GuildMembersChunk, handler);
           this.client.decrementMaxListeners();
           resolve(users && !Array.isArray(users) && fetchedMembers.size ? fetchedMembers.first() : fetchedMembers);
         }
       };
 
-      const timeout = setTimeout(() => {
-        this.client.removeListener(Events.GuildMembersChunk, handler);
-        this.client.decrementMaxListeners();
-        reject(new DiscordjsError(ErrorCodes.GuildMembersTimeout));
-      }, time).unref();
+      const rateLimitHandler = payload => {
+        if (payload.t === GatewayDispatchEvents.RateLimited && payload.d.meta.nonce === nonce) {
+          clearTimeout(timeout);
+          this.client.ws.off(WebSocketShardEvents.Dispatch, rateLimitHandler);
+          this.client.removeListener(Events.GuildMembersChunk, handler);
+          this.client.decrementMaxListeners();
+          reject(new DiscordjsError(ErrorCodes.GatewayRequestRateLimited, payload.d));
+        }
+      };
+
+      this.client.ws.on(WebSocketShardEvents.Dispatch, rateLimitHandler);
```

Comment:
Shouldn't there be a `/src/client/websocket/handlers/RATE_LIMITED.js` for this websocket event so that
- end users can listen to it directly and
- you can use it here directly instead of having to filter the full dispatch event again 

Same applies for the backport where it would replace the usage of listening to `Events.Raw`